pwmig/gclfield2vtk/gclfield2vtk.1
.TH GCLFIELD2VTK 1
.SH NAME
gclfield2vtk \- convert internal gclfield objects to vtk format files
.SH SYNOPSIS
.nf
\fBgclfield2vtk\fR db outfile [-g gridname -f fieldname -r remapgridname 
                -odbf outfieldname -xml|-binary -pf pffile] -V
.fi
.SH DESCRIPTION
.LP
This program is used to convert Geographical CurviLinear (GCL) field objects
stored in an Antelope database to external formats that can be read by 
the Visualization Tool Kit (VTK).  The program can convert scalar3d, vector3d,
scalar2d, and grid2d (structured grid surfaces such as a DEM) objects.  
Outputs are in "legacy VTK" format or the newer XML format depending on the 
choice of options.  Not all combinations are currently supported.  
(see Bugs and Caveats section).
.LP
The \fBdb\fR parameter is required.  It must point to a valid Antelope database
descriptor.  \fBoutfile\fR is the (required) output file name.  Use a simple 
name without an extension for this file.  The program appends an appropriate 
extension depending on the output option.  This is necessary for paraview to
grok the file type in windows, macos, or linux implementations.  The output 
files should normally be readable directly by paraview. 
.SH OPTIONS
.IP -g 
Override the parameter file grid name to locate the gclfield to be converted.
The current implementation stores GCLfield objects in an antelope database
keyed by a gridname and a fieldname.  This is one key.
.IP -f 
Defines the related key for the field name used to index the desired 
data.
.IP -r
Because visualizations use a Cartesian coordinate system it is frequently
essential to make sure a set of outputs are in a common reference frame. 
We commonly do this by defining one grid as a master.  Use the -r option 
to make sure all conversions use the coordinate system of this master.
.IP -odbf
This is most commonly used to build apply an internal agc operator to 
a field and save the result.  This parameter should be avoided an 
will probably eventually be depricated.
.IP -xml
Ue the -xml to have the output written in vtk's xml format.
.IP -binary
Use the -binary format to have the output written as binary.  The program
should automatically test byte order and flag the file appropriately.
.IP -pf
Use pffile.pf as the alternative parameter file to the standard gclfield2vtk.pf.
.SH FILES
.LP
The output file is a legacy VTK file described in documentation for the Visualization Tool Kit (VTK) library.
.SH PARAMETER FILE
.LP
Several parameters can be viewed as defaults for command line arguments.  
The \fBgridname\fR parameter is the default for -g, 
\fBfieldname\fR provides the default for -f, and 
\fBremapgridname\fR provides the default for -r. 
Related is the \fBremap_grid\fR boolean.  If false (the default) \fBremapgridname\fR is ignored 
unless the -r command line argument is used.  Using the -r command line options implicitly turns on 
the remap option.
.LP
The \fBfieldtype\fR defines the type of grid object the program will convert.  Allowed 
options are:  grid2d, scalar2d, scalar3d, and vector3d.  Note that vector3d currently produces
a sequence of scalar field objects for with each vector component in a different file.
.LP
The \fBremapgrid_number_dimensions\fR parameter is used to sort out a potential ambiguity in 
the grid geometry used when remapping is enabled.  That is, the GCLgrid library allows both 2d 
and 3d structured grids.  This assure the coordinate system to use for mapping is unambiguous.
.LP
The boolean parameter \fBsave_as_vector_field\fR is a switch for handling vector field data.
When true output is VTK's vector field format.  When false vector fields are output with one 
file per vector component.  The default is false.  (This capability is not currently implemented.
If set true an error message will be posted and the program will abort.)
.LP
There are some special parameters used only in certain circumstances.  The user is warned these
may one day be depricated in favor of editing and saving to new loction.  The boolean
parameter \fBremove_mean_x3_slices\fR can be useful for absolute velocity tomography models.  
It removes the average value of each layer (constant x3 slice) before saving the output.
The boolean \fBapply_agc\fR and integer valued parameter \fBagc_operator_length\fR are linked.  
When \fBapply_agc\fR is true an agc operator with a length (in vertical-x3 samples) is applied
before output.  The author has used this frequently for wavefield imaging volumes.
.SH DIAGNOSTICS
.LP
Check stderr for errors.  Most problems will generate a relatively verbose message.
.SH "BUGS AND CAVEATS"
.IP (1)
Not all combinations of output format options are support.  The -xml and 
-binary options will only work on 3D scalar or vector fields.  Either
option will silently be ignored for any 2D object.
.IP (2)
Vector fields are not actually properly supported in this version.  As noted
above vector fields are actually output as multiple files that are actually
scalar fields.  VTK supports vector fields, but an output writer has not
yet been implemented.  
.IP (3)
For version 3.8 of paraview on 64 bit platforms a bug seems to exist in libc for reading ascii data created by this program.  The VTK reader for ascii files 
will abort when trying to read numbers with exponents that define tiny
numbers that overflow the exponent field on floats.  (e.g. a number like
1.6698897e-258 will cause the VTK reader for legacy VTK files to throw
an exception and cause paraview, which uses this routine, to abort).  
For this reason use only the -xml or -binary option on 64 bit platforms. 
.IP (4)
This program is a bit of an abomination in that as it evolved it developed a scitzo approach to 
parameter information.  The command line arguments were added to allow easier interaction with
shell scripts but create an inconsistent and (admittedly) confusing parameter definition.
.IP (5)
The agc and remove mean parameters really do not belong in this program.  There is a good chance these
will be replaced with a more elaborate field editor or implemented as a python module in paraview in
the future.
.SH AUTHOR
.nf
Gary L. Pavlis
Department of Geological Sciences
Indiana University 
1001 East 10th Street
Bloomington, IN 47405
pavlis@indiana.edu
.fi


pwmig/gclfield2vtk/gclfield2vtk.cc
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <stdio.h>
#include <float.h>
#include "stock.h"
#include "elog.h"
#include "pf.h"
#include "seispp.h"
#include "Metadata.h"
#include "gclgrid.h"
#include "agc.h"
#include "vtk_output.h"

using namespace SEISPP;

int vtk_output_GCLgrid(GCLgrid& g, string ofile);
/* Removes mean for constant x3 slices.  Important for
proper display of tomography models showing absolute velocities */
void remove_mean_x3(GCLscalarfield3d& f)
{
	double mean;
	int i,j,k;
	double nval=static_cast<double>((f.n1)*(f.n2));
	cout << "Mean removal values"<<endl
		<< "grid-index     mean"<<endl;
	for(k=0;k<f.n3;++k)
	{
		mean=0.0;
		for(i=0;i<f.n1;++i)
			for(j=0;j<f.n2;++j)
				mean += f.val[i][j][k];
		mean /= nval;
		cout << k << "      "<<mean<<endl;
		for(i=0;i<f.n1;++i)
			for(j=0;j<f.n2;++j)
				f.val[i][j][k] = f.val[i][j][k]-mean;
	}
}
/* Applies an agc operator with window length iwagc to data along the x3 direction
then normalizes a vector constructed from each x3 grid line to have rms=1 (L2 norm unity).
*/
void agc_scalar_field(GCLscalarfield3d& g, int iwagc)
{
	vector<double> x3vals;
	double nrmx3v;
	int i,j,k;
	for(i=0;i<g.n1;++i)
		for(j=0;j<g.n2;++j)
		{
			x3vals.clear();
			for(k=0;k<g.n3;++k) x3vals.push_back(g.val[i][j][k]);
			x3vals=agc<double>(x3vals,iwagc);
			nrmx3v=dnrm2(g.n3,&(x3vals[0]),1);
			/* do nothing if nrm is 0 */
			if(nrmx3v<DBL_EPSILON) nrmx3v=1.0;
			for(k=0;k<g.n3;++k) g.val[i][j][k]=x3vals[k]/nrmx3v;
		}
}
		

void usage()
{
	cerr << "gclfield2vtk db outfile [-g gridname -f fieldname -r remapgridname "
		<< "-odbf outfieldname -xml -binary -pf pffile] -V" << endl;
	exit(-1);
}
bool SEISPP::SEISPP_verbose(true);
int main(int argc, char **argv)
{
	int i;

        ios::sync_with_stdio();
	elog_init(argc,argv);
	if(argc<3) usage();
	string dbname(argv[1]);
	string outfile(argv[2]);
	string argstr;
	string pffile("gclfield2vtk");
	const string nodef("NOT DEFINED");
	string gridname(nodef);
	string fieldname(nodef);
	string remapgridname(nodef);
	bool saveagcfield(false);
	string outfieldname;
	bool remap(false);
	bool xmloutput(false);
	bool binaryout(false);
	for(i=3;i<argc;++i)
	{
		argstr=string(argv[i]);
		if(argstr=="-pf")
		{
			++i;
			if(i>=argc)usage();
			pffile=string(argv[i]);
		}
		else if(argstr=="-V")
		{
		    cbanner("1.0",
			"gclfield2vtk db outfile [-pf pffile]",
			"Gary L. Pavlis",
			"Indiana University",
			"pavlis@indiana.edu");
		}
		else if(argstr=="-g")
		{
			++i;
			if(i>=argc)usage();
			gridname=string(argv[i]);
		}
		else if(argstr=="-f")
		{
			++i;
			if(i>=argc)usage();
			fieldname=string(argv[i]);
		}
		else if(argstr=="-r")
		{
			++i;
			if(i>=argc)usage();
			remapgridname=string(argv[i]);
			remap=true;
		}
		else if(argstr=="-odbf")
		{
			++i;
			if(i>=argc)usage();
			saveagcfield=true;
			outfieldname=string(argv[i]);
		}
		else if(argstr=="-xml")
		{
			xmloutput=true;
		}
		else if(argstr=="-binary")
		{
			binaryout=true;
		}
		else
		{
			usage();
		}
	}
        Pf *pf;
        if(pfread(const_cast<char *>(pffile.c_str()),&pf))
                elog_die(1,"pfread error\n");

	try {
        	Metadata control(pf);
                DatascopeHandle dbh(dbname,true);

		if(gridname==nodef)
			gridname=control.get_string("gridname");
		if(fieldname==nodef)
			fieldname=control.get_string("fieldname");
		string fieldtype=control.get_string("fieldtype");
		cout << "Converting field = "<<fieldname
			<< " associated with gridname="<<gridname<<endl;
		cout << "Assuming grid type is "<<fieldtype<<endl;
		/* Slightly odd logic here, but this allows remap off
		to be the default.  pf switch is ignored this way if
		the -r flag was used */
		if(!remap) remap=control.get_bool("remap_grid");
		BasicGCLgrid *rgptr;
		if(remap)
		{
			if(remapgridname==nodef)
				remapgridname=control.get_string("remapgridname");
			cout << "Remapping emabled.  "
				<<"Will use coordinate system of grid "
				<< remapgridname<<endl;
			int griddim=control.get_int("remapgrid_number_dimensions");
			if(griddim==3)
			{
				GCLgrid3d *gtmp=new GCLgrid3d(dbh,
					remapgridname);
				rgptr=dynamic_cast<BasicGCLgrid*>(gtmp);
			}
			else if(griddim==2)
			{
				GCLgrid *gtmp=new GCLgrid(dbh,
					remapgridname);
				rgptr=dynamic_cast<BasicGCLgrid*>(gtmp);
			}
			else
			{
				cerr << "Illegal remap grid dimension="
					<<griddim<<endl
					<<"Fix parameter file"<<endl;
			}
		}
			
		bool SaveAsVectorField;
		int VectorComponent;
		bool rmeanx3=control.get_bool("remove_mean_x3_slices");
		bool apply_agc=control.get_bool("apply_agc");
		int iwagc(0);
		if(apply_agc)
		{
			iwagc=control.get_int("agc_operator_length");
		}
		if(saveagcfield  && !apply_agc)
		{
			cerr << "Illegal option combination"<<endl
			  << "-odbf argument can only be used if apply_agc"
			  << " pf parameter is set true"<<endl;
			exit(-1);
		}
		string fielddir;
		if(saveagcfield)
			fielddir=control.get_string("field_directory");
		if(fieldtype=="scalar3d") 
		{
			GCLscalarfield3d field(dbh,gridname,fieldname);
			if(remap)
			{
				// Used to make this optional.  force
				//if(field!=(*rgptr))
					remap_grid(dynamic_cast<GCLgrid3d&>(field),
						*rgptr);
			}
			if(rmeanx3) remove_mean_x3(field);
			if(apply_agc) agc_scalar_field(field,iwagc);
			output_gcl3d_to_vtksg(field,outfile,xmloutput,binaryout);
			if(saveagcfield) 
				field.save(dbh,string(""),fielddir,
				  outfieldname,outfieldname);
		}
		else if(fieldtype=="vector3d")
		{
			SaveAsVectorField=control.get_bool("save_as_vector_field");
			if(SaveAsVectorField)
			{
				cerr << "Vector field output not yet supported but planned"
					<< endl;
			}
			if(rmeanx3)
			{
				cerr << "remove_mean_x3_slices set true:  "
					<< "ignored for vector field="
					<< fieldname<<endl;
			}
                        /* This needs a more general solution long term.  For now this is
                           frozen for vector size in pwmig */
                        const int nvpwmig(5);
			GCLvectorfield3d vfield(dbh,gridname,fieldname,nvpwmig);
			GCLscalarfield3d *sfptr;
			if(remap)
			{
				//if(vfield!=(*rgptr))
					remap_grid(dynamic_cast<GCLgrid3d&>(vfield),
						*rgptr);
			}
			for(i=0;i<vfield.nv;++i)
			{
				sfptr = extract_component(vfield,i);
				if(apply_agc) agc_scalar_field(*sfptr,iwagc);
				stringstream ss;
				ss << outfile <<"_"<<i;
				if(xmloutput) 
					ss<<".vts";
				else
					ss<<".vtk";
				output_gcl3d_to_vtksg(*sfptr,
					ss.str(),xmloutput,binaryout);
				/*This is not ideal, but will do this now
				for expedience.  This creates a series of 
				scalar fields when saveagcfield is enabled
				that have a naming convention similar to
				the vtk files.  These could easily overflow
				but I don't test them here. beware */
				if(saveagcfield) 
				{
					stringstream ssof;
					ssof << outfieldname<<"_"<<i;
					string ofld=ssof.str();
					sfptr->save(dbh,string(""),
					  fielddir,ofld,ofld);
				}
				delete sfptr;
			}
		}
		else if(fieldtype=="grid2d")
		{
			int npoly;
			GCLgrid g(dbh,gridname);
			if(remap)
			{
				//if(g!=(*rgptr))
					remap_grid(g,*rgptr);
			}
			npoly=vtk_output_GCLgrid(g,outfile);
			cout << "Wrote "<<npoly<<" polygons to output file"<<endl;
			if(apply_agc)cerr <<"WARNING: apply_agc was set true\n"
					<<"This is ignored for grids and 2d fields"<<endl;
		}
		else if(fieldtype=="scalar2d")
		{
			int npoly;
			GCLscalarfield field(dbh,gridname,fieldname);
			if(remap)
			{
				//if(field!=(*rgptr))
					remap_grid(dynamic_cast<GCLgrid&>(field),
						*rgptr);
			}
			vtk_output_GCLgrid(dynamic_cast<GCLgrid&>(field),outfile);
			cout << "Wrote "<<npoly<<" polygons to output file"<<endl;
			if(apply_agc)cerr <<"WARNING: apply_agc was set true\n"
					<<"This is ignored for grids and 2d fields"<<endl;
		}	
		else
		{
			cerr << "Unsupported fieldtype = "<<fieldtype<<endl
				<< "Exiting with no output\n" << endl;
		}

	}
	catch (MetadataGetError mderr)
	{
		mderr.log_error();
		exit(-1);
	}
	catch(int ierr)
	{
		elog_die(1,"GCLgrid library error\n");
	}
	catch (...)
	{
		elog_die(1,"Something threw and unhandled excepton\n");
	}
}
